tuna/gen_fastdb.py

```diff
@@ -0,0 +1,298 @@
+                                          ['solver', 'kernel_time'])
+      duplicates = duplicates.groupby(cols_with_conv_params + ['solver'])
+
+      duplicates_str = as_heading("Duplicates") + '\n'
+      for i, (_, df) in enumerate(duplicates):  # pylint: disable=invalid-name
+        for _, row in df.iterrows():
+          conv_params_str = "+ " if i % 2 == 0 else "- "
+          for colname, val in zip(cols_with_conv_params,
+                                  row[cols_with_conv_params]):
+            conv_params_str += f'{colname} {val}, '
+          duplicates_str += f"{conv_params_str}\n  solver: {row['solver']},  " +\
+                    f"kernel_time: {row['kernel_time']}\n\n"
+
+      logging.log(duplicates_str, silent=True)
+
+    if strict:
+      raise AssertionError(
+          f'{num_duplicates} duplicate entries detected in {tag}finddb')
+    logging.warning(
+        f'{num_duplicates} duplicate entries delected in {tag}finddb')
+    issues.append(DUPLICATE_ENTRIES)
+
+  if not issues:
+    logging.success(f'{tag}finddb passed all checks!')
+
+  return issues
+
+
+def describe_fastdb(fastdb, tag=''):
+  """ describe fastdb (num. of entries in it, etc.) """
+  if len(fastdb) == 0:
+    logging.warning(f'{tag}fastdb empty!')
+  else:
+    logging.info(
+        f'Total entries in {tag}fastdb (=unique configs in {tag}finddb): %d' %
+        len(fastdb))
+    logging.info(f'Number of unique solvers in {tag}fastdb: %d' %
+                 len(fastdb['solver'].unique()))
+
+
+def load_fastdb(fastdb_pickle_filename):
+  """ load fastdb from pickle """
+  fastdb = io_tools.safe_load(None, fastdb_pickle_filename,
+                              df_tools.from_pickle)
+  describe_fastdb(fastdb)
+  return fastdb
+
+
+def finddb_to_nonthresholded_fastdb(finddb,
+                                    cols_with_conv_params=None,
+                                    n_fastest=1,
+                                    tag=''):
+  """ generate fastdb from finddb without thresholding/removing out any solvers """
+  if cols_with_conv_params is None:
+    cols_with_conv_params = _DEFAULT_COLS_WITH_CONV_PARAMS
+
+  num_duplicates = finddb.duplicated(subset=cols_with_conv_params +
+                                     ['solver']).sum()
+  if num_duplicates > 0:
+    logging.warning(f'resolved the {num_duplicates} duplicate (fdb_key, solver)' +\
+            f' pairs found in {tag}finddb by keeping the entries with fastest solver')
+  sorted_finddb = finddb.sort_values(cols_with_conv_params + ['kernel_time'])
+  return sorted_finddb.groupby(cols_with_conv_params).head(n_fastest)
+
+
+def is_fundamental_solver(solver):
+  """defines fundamental solvers"""
+
+  def is_explicit_gemm(solver):
+    return is_substr('Gemm', solver) and not is_substr('ImplicitGemm', solver)
+
+  def is_naive(solver):
+    return is_substr('Naive', solver)
+
+  return is_explicit_gemm(solver) or is_naive(solver)
+
+
+# pylint: disable=too-many-locals, too-many-branches
```

Comment:
same here, please do not use these disable flags, we are using pylint to explicitly enforce addressing issues such as too-many-locals and too-many-branches